zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,362 @@
+                    // invariant from `p`, so we re-assert that all of the bytes
+                    // are initialized.
+                    let variant = unsafe { variant.assume_initialized() };
+                    <
+                        #variant_struct_ident #ty_generics as #trait_path
+                    >::is_bit_valid(variant)
+                }
+            }
+        }
+    });
+
+    quote! {
+        // SAFETY: We use `is_bit_valid` to validate that the bit pattern of the
+        // enum's tag corresponds to one of the enum's discriminants. Then, we
+        // check the bit validity of each field of the corresponding variant.
+        // Thus, this is a sound implementation of `is_bit_valid`.
+        fn is_bit_valid<A>(
+            mut candidate: ::zerocopy::Maybe<'_, Self, A>,
+        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool
+        where
+            A: ::zerocopy::pointer::invariant::Aliasing
+                + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>,
+        {
+            use ::zerocopy::macro_util::core_reexport;
+
+            #tag_enum
+
+            type ___ZerocopyTagPrimitive = ::zerocopy::macro_util::SizeToTag<{
+                core_reexport::mem::size_of::<___ZerocopyTag>()
+            }>;
+
+            #tag_consts
+
+            type ___ZerocopyOuterTag = #outer_tag_type;
+            type ___ZerocopyInnerTag = #inner_tag_type;
+
+            #variant_structs
+
+            #variants_union
+
+            #[repr(C)]
+            struct ___ZerocopyRawEnum #generics {
+                tag: ___ZerocopyOuterTag,
+                variants: ___ZerocopyVariants #ty_generics,
+            }
+
+            let tag = {
+                // SAFETY:
+                // - The provided cast addresses a subset of the bytes addressed
+                //   by `candidate` because it addresses the starting tag of the
+                //   enum. Because the pointer is cast from `candidate`, it has
+                //   the same provenance as it. There are no `UnsafeCell`s in
+                //   the tag because it is a primitive integer.
+                // - By definition, `Self` and `__ZerocopyTagPrimitive` are
+                //   types of the same size.
+                let tag_ptr = unsafe {
+                    candidate.reborrow().cast_unsized(|p: *mut Self| {
+                        p as *mut ___ZerocopyTagPrimitive
+                    })
+                };
+                // SAFETY: `candidate` is initialized, and `tag_ptr` points to
+                // a region completely contained within `candidate`. Therefore,
+                // `tag_ptr` is also initialized.
+                let tag_ptr = unsafe { tag_ptr.assume_initialized() };
+                tag_ptr.bikeshed_recall_valid().read_unaligned()
+            };
+
+            // SAFETY:
+            // - The raw enum has the same size as the input enum.
```

Comment:
The derive needs to forward `align` annotations where appropriate. For example, deriving `TryFromBytes` on this:
```rust
#[repr(C, align(64))]
enum MyEnum {
    Foo(u8),
}
```
violates this safety condition. `___ZerocopyRawEnum` has a size of `8` on x86_64, whereas `MyEnum` has a size of `64`.

Reply:
That isn't be required for safety. The alignment and overall size of the original enum is higher, but we don't use the raw enum for anything except projecting into the fields. I have reworded this safety comment to note that the raw enum may have lower alignment and size.